frontends/concrete-python/concrete/fhe/extensions/synthesis/verilog_to_luts.py

```diff
@@ -0,0 +1,346 @@
+        print()
+    if fatal:
+        raise ValueError(fatal) from None
+
+
+def yosys_run_script(verilog, yosys_dot_file):
+    """Run the yosys script with the adequate generated files."""
+    tmpdir_prefix = Path.home() / ".cache" / "concrete-python" / "synthesis"
+    os.makedirs(tmpdir_prefix, exist_ok=True)
+
+    def tmpfile(mode, suffix, **kwargs):
+        return tempfile.NamedTemporaryFile(
+            mode=mode, dir=tmpdir_prefix, suffix=f"-{suffix}", **kwargs
+        )
+
+    # fmt: off
+    with \
+            tmpfile("w+", "script.abc") as abc_file, \
+            tmpfile("w+", "lut-costs.txt") as lut_costs_file, \
+            tmpfile("w+", "source.verilog") as verilog_file, \
+            tmpfile("w+", "yosys.script") as yosys_file, \
+            tmpfile("r", "luts.json", delete=False) as json_file:
+        return _yosys_run_script(
+            abc_file, lut_costs_file, yosys_file,
+            verilog_file, verilog,
+            json_file, dot_file=yosys_dot_file,
+        )
+    # fmt: on
+
+
+@dataclass(frozen=True)
+class ValueOrigin:
+    """Original verilog input/output information for a value."""
+
+    is_parameter: bool = False
+    is_result: bool = False
+    bit_index: int = 0
+
+
+@dataclass
+class ValueNode:
+    """An intermediate named value."""
+
+    name: str
+    origin: ValueOrigin = ValueOrigin()
+    base_name: Optional[str] = None
+
+    @classmethod
+    def parameter(cls, base_name, i):
+        """Construct a parameter (i.e. verilog input) derived value from the base input name."""
+        # E.g. `ValueNode.parameter("a", 3)` is "a[3]", the 4th bit of verilog input `a`.
+        return ValueNode(
+            f"{base_name}[{i}]", ValueOrigin(is_parameter=True, bit_index=i), base_name
+        )
+
+    @classmethod
+    def result(cls, base_name, i):
+        """Construct a result (i.e. verilog input) derived value from the base output name."""
+        # E.g. `ValueNode.result("r", 3)` is "r[3]", the 4th bit of verilog output `r`.
+        return ValueNode(f"{base_name}[{i}]", ValueOrigin(is_result=True, bit_index=i), base_name)
+
+    @classmethod
+    def port(cls, base_name, port, i):
+        """Construct a verilog port, i.e. either parameter/input or result/output."""
+        if port["direction"] == "input":
+            return cls.parameter(base_name, i)
+        assert port["direction"] == "output"
+        return cls.result(base_name, i)
+
+    @property
+    def is_interface(self):
+        """Check if a value is part of the verilog circuit interface."""
+        return self.origin.is_parameter or self.origin.is_result
+
+
+@dataclass
+class TluNode:
+    """A TLU operation node."""
+
+    arguments: List[ValueNode]
+    results: List[ValueNode]
+    content: List[Any]
+
+    @property
+    def arity(self):
+        """Number of single bit parameters of the TLU."""
+        return len(self.arguments)
+
+
+@dataclass
+class Circuit:
```

Comment:
Please avoid duplication of existing class names. Let's go for `VerilogCircuit` or something like that.